Support multiple passengers

Postremus opened the pull request:
This PR attempts to implement multiple passengers per vehicle.

Save and load Passengers. When an entity is a passenger, it will only be saved as child of its vehicle, and never alone.
Adjust api implementation to be (hopefully) in line with papers.
Also set the size for armorstands and horses.

closes #522
related to #504

Postremus committed:
Support multiple passengers

save and load Passengers
adjust api implementation to be (hopefully) in line with paper's
set size for armorstands and horses

closes #522
related to #504

Postremus committed:
Catch exceptions when trying to load a passenger

Postremus committed:
Teleport passengers to their vehicle
set size of minecart

aramperes commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -1105,73 +1109,92 @@ public void setCustomNameVisible(boolean flag) {
     @Override
     public boolean removePassenger(Entity passenger) {
-        return passenger.leaveVehicle();
+        Preconditions.checkArgument(passenger != this, "Entity cannot ride itself.");
```
Replace `passenger != this` with `!Objects.equals(passenger, this)`

aramperes commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -1105,73 +1109,92 @@ public void setCustomNameVisible(boolean flag) {
     public boolean addPassenger(Entity passenger) {
-        return setPassenger(passenger); // TODO: multiple passengers
-    }
+        Preconditions.checkArgument(passenger != this, "Entity cannot ride itself.");
```
See previous review comment (don't use `!=`/`==` for equality checks)

aramperes commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -1105,73 +1109,92 @@ public void setCustomNameVisible(boolean flag) {
 
-            GlowEntity passenger = (GlowEntity) bPassenger;
+        Location onTopOfVehicle = this.location.clone().add(0, this.getHeight(), 0);
+        glowPassenger.teleport(onTopOfVehicle);
```
I think you would also need to teleport the passenger after dismounting, no?

aramperes commented on src/main/java/net/glowstone/io/entity/EntityStore.java:
```diff
@@ -98,6 +102,37 @@ public void load(T entity, CompoundTag tag) {
+            for (CompoundTag entityTag : tag.getCompoundList("Passengers")) {
+                Entity passenger = loadPassenger(entity, entityTag);
+                if (passenger != null) {
+                    entity.addPassenger(passenger);
```
Perhaps you should make another ("internal") method when setting the passengers from NBT -- you wouldn't want to fire these events when loading the world.

Postremus replied:
I have to agree with you, that you don't want to do this normally. Spigot however does throw them during world load..

aramperes replied:
Ah okay, that's fine then.

Postremus committed:
Also teleport passengers on dismount
dead entities do not have any passengers

Mother, Mother, why is the cow hovering?

aramperes commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -898,6 +899,9 @@ protected void pulsePhysics() {
             setRawLocation(location.clone().add(velocity));
         }
 
+        // apply location changes to all passengers
```
Empty comment block? Or is there something missing here?

aramperes commented on src/main/java/net/glowstone/io/entity/EntityStore.java:
```diff
@@ -98,6 +102,37 @@ public void load(T entity, CompoundTag tag) {
+            return EntityStorage.loadEntity(vehicle.getWorld(), compoundTag);
+        } catch (Exception e) {
+            String id = compoundTag.isString("id") ? compoundTag.getString("id") : "<missing>";
+            GlowServer.logger.warning("Skipping Entity with id " + id);
```
Log the exception here

Postremus committed:
log exception

Postremus committed:
fix whitespace

aramperes commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -134,7 +135,7 @@
      * Passenger
      */
-    private GlowEntity passenger;
+    private List<Entity> passengers = new ArrayList<>();
```
This field can be final (and perhaps `protected` too)

Postremus committed:
make passengers list final
